bpf/lib/nat.h

```diff
@@ -1108,7 +1101,14 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target,
 		case ICMP_DEST_UNREACH:
 			if (icmphdr.code != ICMP_FRAG_NEEDED)
 				return NAT_PUNT_TO_STACK;
-			return snat_v4_rev_nat_handle_icmp_frag_needed(ctx, off);
+
+			ret = snat_v4_rev_nat_handle_icmp_frag_needed(ctx, off, &state);
+			if (IS_ERR(ret))
+				return ret;
+
+			/* Skip L4 rewrite: */
+			tuple.dport = state->to_dport;
```

Comment:
The're no port or such in the outer `ICMP_FRAG_NEEDED` header that we need to rewrite. Right now `snat_v4_rewrite_ingress()` would access the ICMP header again, determine the type, and skip the processing (since it's not `ICMP_ECHOREPLY`. So we can also avoid the header access in this case.

But you're right, it's not exactly equivalent. And worth its own patch I suppose. So let me back it out.

Reply:
> Did you move it to some other pull request?

I'll pick it up (or something equivalent) with a later PR, yep. Still plenty of NAT rework to do :grin: .